log4j-api/src/main/java/org/apache/logging/log4j/spi/internal/DefaultScopedContextProvider.java

```diff
@@ -0,0 +1,421 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.spi.internal;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.function.Supplier;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.ScopedContext;
+import org.apache.logging.log4j.ThreadContext;
+import org.apache.logging.log4j.spi.ScopedContextProvider;
+import org.apache.logging.log4j.status.StatusLogger;
+import org.apache.logging.log4j.util.StringMap;
+
+/**
+ * An implementation of {@link ScopedContextProvider} that uses the simplest implementation.
+ * @since 2.24.0
+ */
+public class DefaultScopedContextProvider implements ScopedContextProvider {
+
+    public static final Logger LOGGER = StatusLogger.getLogger();
+    public static final ScopedContextProvider INSTANCE = new DefaultScopedContextProvider();
+
+    private final ThreadLocal<Instance> scopedContext = new ThreadLocal<>();
+
+    /**
+     * Returns an immutable Map containing all the key/value pairs as Object objects.
+     * @return The current context Instance.
+     */
+    private Optional<Instance> getContext() {
+        return Optional.ofNullable(scopedContext.get());
```

Comment:
This might allocate an instance of `Optional` at each call. Maybe we could return a constant:

```java
private final ScopedContext.Instance EMPTY_INSTANCE = new Instance(this, Collections.emptyMap());
```

if nothing is attached to the `ThreadLocal`?

Reply:
This suggestion is included in the #2450 proposal.

Reply:
Actually, it will return a static empty Optional if the value is null. Returning an empty instance just confuses things.